pubsublite/internal/wire/acks.go

```diff
@@ -0,0 +1,241 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+
+package wire
+
+import (
+	"container/list"
+	"fmt"
+	"sync"
+)
+
+// AckConsumer is the interface exported from this package for acking messages.
+type AckConsumer interface {
+	Ack()
+}
+
+// ackedFunc is invoked when a message has been acked by the user.
+type ackedFunc func(*ackConsumer)
+
+// ackConsumer is used for handling message acks. It is attached to a Message
+// and also stored within the ackTracker until the message has been acked by the
+// user.
+type ackConsumer struct {
+	// The message offset.
+	Offset int64
+	// Bytes released to the flow controller once the message has been acked.
+	MsgBytes int64
+
+	// Guards access to fields below.
+	mu    sync.Mutex
+	acked bool
+	onAck ackedFunc
+}
+
+func newAckConsumer(offset, msgBytes int64, onAck ackedFunc) *ackConsumer {
+	return &ackConsumer{Offset: offset, MsgBytes: msgBytes, onAck: onAck}
+}
+
+func (ac *ackConsumer) Ack() {
+	ac.mu.Lock()
+	defer ac.mu.Unlock()
+
+	if ac.acked {
+		return
+	}
+	ac.acked = true
+	if ac.onAck != nil {
+		ac.onAck(ac)
```

Comment:
It's a little scary to make arbitrary function calls while holding a lock, since there is nothing preventing the user from attempting a down-call that tries to grab the same lock -- like in your test case. Can you just fire a go-routine here instead?

Reply:
Yes, agreed. As discussed offline, it was originally async here, but complicated the unit tests. The ackFunc implementation fires the goroutine instead. Added a comment to ackedFunc to clarify.